File: apps/web/src/types/ws.ts
/** WebSocket 消息协议类型 (前后端共享契约, 见 apps/api/app/schemas/ws.py) */

import type { ChatMessage } from './chat'

/** 客户端 -> 服务端 */
export interface ClientMessageMap {
  ping: { type: 'ping' }
  join: { type: 'join'; room: string }
  leave: { type: 'leave'; room: string }
  chat: { type: 'chat'; room: string; content: string }
  typing: { type: 'typing'; room: string }
}

export type ClientMessageType = keyof ClientMessageMap
export type ClientMessage = ClientMessageMap[ClientMessageType]

/** WebSocket 中的用户摘要 (welcome 消息) */
export interface WsUserSummary {
  id: number
  username: string
  is_superuser: boolean
}

/** 服务端 -> 客户端 (联合类型, 用 type 判别) */
export type ServerMessage =
  | { type: 'pong'; ts?: number }
  | { type: 'welcome'; ts?: number; user: WsUserSummary; rooms: string[] }
  | { type: 'joined'; room: string; ts?: number }
  | { type: 'left'; room: string; ts?: number }
  | { type: 'chat'; room: string; message: ChatMessage }
  | { type: 'presence'; room: string; online: number; users: string[] }
  | { type: 'notification'; level: 'info' | 'success' | 'warning' | 'error'; title: string; content: string; room?: string }
  | { type: 'error'; content: string }

/** WebSocket 连接状态 */
export type WsStatus = 'idle' | 'connecting' | 'open' | 'closing' | 'closed' | 'reconnecting'

File: apps/api/app/schemas/ws.py
"""WebSocket 消息协议 Schema (前后端共享契约)。

客户端 -> 服务端
    {"type": "ping"}
    {"type": "join",  "room": "lobby"}
    {"type": "leave", "room": "lobby"}
    {"type": "chat",  "room": "lobby", "content": "hi"}
    {"type": "typing","room": "lobby"}

服务端 -> 客户端
    {"type": "pong",      "ts": 1710000000}
    {"type": "welcome",   "user": {...}, "rooms": [...]}
    {"type": "joined",    "room": "lobby"}
    {"type": "chat",      "room": "lobby", "message": {...}}
    {"type": "presence",  "room": "lobby", "online": 3, "users": [...]}
    {"type": "notification", "level": "info", "title": "...", "content": "..."}
    {"type": "error",     "message": "..."}
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

ClientMessageType = Literal["ping", "join", "leave", "chat", "typing"]
ServerMessageType = Literal[
    "pong", "welcome", "joined", "left", "chat", "presence", "notification", "error"
]


class ClientMessage(BaseModel):
    type: ClientMessageType
    room: str | None = Field(default=None, max_length=64)
    content: str | None = Field(default=None, max_length=4000)


class ChatMessageOut(BaseModel):
    id: int | None = None
    room: str
    sender_id: int | None = None
    sender_name: str
    kind: str = "text"
    content: str
    created_at: datetime
    source: str | None = None


class ServerMessage(BaseModel):
    type: ServerMessageType
    ts: int | None = None
    room: str | None = None
    message: ChatMessageOut | None = None
    online: int | None = None
    users: list[str] | None = None
    level: Literal["info", "success", "warning", "error"] | None = None
    title: str | None = None
    content: str | None = None
    user: dict[str, Any] | None = None
    rooms: list[str] | None = None

    def dump(self) -> str:
        return self.model_dump_json(exclude_none=True, by_alias=True)


class RoomOut(BaseModel):
    id: int
    code: str
    name: str
    description: str | None = None
    is_default: bool = False
